CORE-19432 - option to disable connection pooling

driessamyn opened the pull request:
- Option to disable connection pooling.
- To validate, disable for Virtual Node reset operation and certificate service.
- Follow-up PRs will further disable other operations.

Commit:
Add option to skip the client side pooling

Commit:
Disable client side pooling for certificate service and vnode reset operation.

Commit:
Fix integration test

yift-r3 commented on components/membership/certificates-service-impl/src/main/kotlin/net/corda/membership/certificate/service/impl/DbClientImpl.kt:
```diff
@@ -102,7 +102,8 @@ internal class DbClientImpl(
                     "persistenceUnitName ${CordaDb.Vault.persistenceUnitName} is not registered."
-                )
+                ),
+            enablePool = false
```
Why don't we want to enable pooling on the virtual node certificate processor?

driessamyn replied:
Because we're removing it from all admin operations to reduce the number of connections in use.
There will still be serverside pooling, so the latency overhead should be minimal.

yift-r3 replied:
The title is a little bit confusing, it's not really "option to disable...", maybe "optionally disable..."?

driessamyn replied:
The description (and commit log) mentions I disabled it for the cert service.

There is also https://github.com/corda/corda-runtime-os/pull/5603, but the reason I added one here was to validate the change works and this one was the smallest change I believe.

simon-johnson-r3 commented on components/db/db-connection-manager-impl/src/main/kotlin/net/corda/db/connection/manager/impl/DbConnectionOpsCachedImpl.kt:
```diff
@@ -64,10 +64,11 @@ class DbConnectionOpsCachedImpl(
     ): EntityManagerFactory {
-        return cacheByConnectionId.computeIfAbsent(connectionId) {
-            delegate.createEntityManagerFactory(connectionId, entitiesSet)
+        return cacheByConnectionId.computeIfAbsent(Pair(connectionId, enablePool)) {
```
this function already had a bug that the cache doesn't take account of entitiesSet, I wonder if you want to fix that at the same time, because adding something else to the cache key just makes that bug even more confusing for users of this method?

driessamyn replied:
As discussed, change the implementation to disallow this case.
Let me know if this makes sense.

simon-johnson-r3 commented on components/db/db-connection-manager/src/main/kotlin/net/corda/db/connection/manager/DbConnectionOps.kt:
```diff
@@ -141,7 +141,20 @@ interface DbConnectionOps {
+    /**
+     * Get an [EntityManagerFactory] for a given connection ID. Use cache or create one if necessary.
+     *
+     * A consumer should not close the EMF.
```
This is a bug waiting to happen, I'd suggest we employ the quite simple solution we already use here https://github.com/corda/corda-runtime-os/blob/c4fb3df8709f1072bf602800873a575502503cbf/components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt#L27 to make the call to close "safe" for things returned from these methods.

driessamyn replied:
Makes sense. Changed.

Commit:
- Make close no-op for EMFs returned from getOrCreateEntityManagerFactory
- Disallow creating EMFs for the same connection ID with a different EntitySet.

Commit:
- Add test for disalllowed operation.

Commit:
- Add test for no-op close operation.

kyriathar commented on components/db/db-connection-manager-impl/src/main/kotlin/net/corda/db/connection/manager/impl/DbConnectionOpsCachedImpl.kt:
```diff
@@ -22,7 +22,7 @@ class DbConnectionOpsCachedImpl(
     private val cache = ConcurrentHashMap<Pair<String,DbPrivilege>, EntityManagerFactory>()
 
-    private val cacheByConnectionId = ConcurrentHashMap<UUID, EntityManagerFactory>()
+    private val cacheByConnectionId = ConcurrentHashMap<Pair<UUID,Boolean>, Pair<EntityManagerFactory,Int>>()
```
Should we make this API a bit more intuitive? I.e. replace `Pair`s with data classes like `ConnectionInfo` (for the first one.)

Second one could be something with id?

driessamyn replied:
The Pairs aren't exposed to the API, so I don't think there's much value in converting this into a data class.

kyriathar commented on components/db/db-connection-manager-impl/src/main/kotlin/net/corda/db/connection/manager/impl/DbConnectionOpsCachedImpl.kt:
```diff
@@ -64,10 +64,19 @@ class DbConnectionOpsCachedImpl(
     ): EntityManagerFactory {
-        return cacheByConnectionId.computeIfAbsent(connectionId) {
-            delegate.createEntityManagerFactory(connectionId, entitiesSet)
+        val entities = entitiesSet.classes.hashCode()
```
Perhaps rename to `entitiesId` ?

driessamyn replied:
It's not really and ID, it's the hashcode for the entitySet ;)

kyriathar commented on components/db/db-connection-manager-impl/src/main/kotlin/net/corda/db/connection/manager/impl/DbConnectionOpsCachedImpl.kt:
```diff
@@ -64,10 +64,19 @@ class DbConnectionOpsCachedImpl(
+        val emfP = cacheByConnectionId.computeIfAbsent(Pair(connectionId, enablePool)) {
+            Pair(delegate.createEntityManagerFactory(connectionId, entitiesSet, enablePool), entities)
+        }
+        if(entities != emfP.second)
+            throw IllegalArgumentException("EntityManagerFactory with a different JpaEntitiesSet already exists.")
```
Did we use to do this?

simon-johnson-r3 replied:
if you remember we discussed some time back it should include this in the cache key, I suggested this could be done now in a previous comment in this review, @driessamyn suggested instead of adding it to the key we consider it a programming error, which I think is reasonable

kyriathar commented on components/db/db-connection-manager/src/main/kotlin/net/corda/db/connection/manager/DbConnectionOps.kt:
```diff
@@ -141,7 +141,20 @@ interface DbConnectionOps {
      * @return
      */
-    fun createEntityManagerFactory(connectionId: UUID, entitiesSet: JpaEntitiesSet): EntityManagerFactory
+    fun createEntityManagerFactory(connectionId: UUID, entitiesSet: JpaEntitiesSet, enablePool: Boolean = true):
```
I can remember this has been discussed in the past (and this is not for this PR), could we at some point make things to be EMF detangled from underlying Datasource? So that then it means one EMF instance per entity set to be shared across multiple datasources pointing to same schema?

@driessamyn @simon-johnson-r3

driessamyn replied:
I don't think that is possible in the EMF model. The datasource is part of the Hibernate EMF.
"Untangling" this may mean quite a lot of customisation, and I'd rather spend the time figuring out how to not use Hibernate where we don't need to.

kyriathar reviewed (commented):
Few comments (mostly nits around API)

yift-r3 reviewed (approved):
I prefer to change the title,

Regardless - the network code LGTM.
